DOCS-4627 Source Code Integration Doc Edits (#16506)
DataDog/documentation: content/en/integrations/guide/source-code-integration.md

---
title: Datadog Source Code Integration
kind: guide

> description: "Set up the source code integration that integrates with APM to link your telemetry with your backend services' repositories and generate inline code snippets."

Author: ```suggestion
description: "Set up the source code integration that integrates with APM to link your telemetry with your  repositories and generate inline code snippets."
```

further_reading:
- link: "https://docs.datadoghq.com/integrations/github/"
  tag: "Documentation"
  text: "Learn about the GitHub Integration"
- link: "https://docs.datadoghq.com/tracing/error_tracking/"
  tag: "Documentation"
  text: "Learn about Error Tracking for Backend Services"
- link: "https://docs.datadoghq.com/profiler/"
  tag: "Documentation"
  text: "Learn about the Continuous Profiler"
---

## Overview

<div class="alert alert-info">

> The source code integration supports the following languages (Go, Java, JavaScript, and Python) and the following Git providers (GitHub, GitLab, and BitBucket).</br></br> Custom URLs are not supported—while your Git provider may be GitHub, if your source code is hosted on <code>xyz.com</code>, for example, and not <code>github.com</code>, the source code integration does not work.

Reviewer: ```suggestion
The source code integration supports:</br>Languages:<ul><li>Go</li><li>Java</li><li>JavaScript</li><li>Python</li></ul></br>Git providers:<ul><li>GitHub</li><li>GitLab</li><li>BitBucket</li></br></br> Custom URLs are not supported—while your Git provider may be GitHub, if your source code is hosted on <code>xyz.com</code>, for example, and not <code>github.com</code>, the source code integration does not work.
```
I wonder if unordered lists would be more natural here than "the following" + parentheses?

> The source code integration supports:</br></br>Languages:<ul><li>Go</li><li>Java</li><li>JavaScript</li><li>Python</li></ul></br>Git providers:<ul><li>GitHub</li><li>GitLab</li><li>BitBucket</li></br> Custom URLs are not supported—while your Git provider may be GitHub, if your source code is hosted on <code>xyz.com</code>, for example, and not <code>github.com</code>, the source code integration does not work.

Reviewer: ```suggestion
The source code integration supports:</br></br>Languages:<ul><li>Go</li><li>Java</li><li>JavaScript</li><li>Python</li></ul></br>Git providers:<ul><li>GitHub</li><li>GitLab</li><li>BitBucket</li></ul></br> Custom URLs are not supported—while your Git provider may be GitHub, if your source code is hosted on <code>xyz.com</code>, for example, and not <code>github.com</code>, the source code integration does not work.
```

> The source code integration supports:</br></br>Languages:<ul><li>Go</li><li>Java</li><li>JavaScript</li><li>Python</li></ul></br>Git providers:<ul><li>GitHub</li><li>GitLab</li><li>BitBucket</li></ul></br> Custom URLs are not supported—while your Git provider may be GitHub, if your source code is hosted on <code>xyz.com</code>, for example, and not <code>github.com</code>, the source code integration does not work.

Author: ```suggestion
The source code integration supports:</br></br>Languages:<ul><li>Go</li><li>Java</li><li>JavaScript (doesn't support transpired JavaScript)</li><li>Python</li></ul></br>Git providers:<ul><li>GitHub</li><li>GitLab</li><li>BitBucket</li></ul></br> Self-hosted instances or private URLs are not supported.
```

</div>

> Datadog's source code integration allows you to connect your telemetry with your Git repositories hosted in GitHub, GitLab, or Bitbucket. Once you have enabled the [GitHub integration][5], you can debug stack traces, slow profiles, and more issues by quickly accessing the relevant lines of your source code.

Reviewer: ```suggestion
Datadog's source code integration allows you to connect your telemetry with your Git repositories hosted in GitHub, GitLab, or Bitbucket. Once you have enabled the [GitHub integration][5], you can debug stack traces, slow profiles, and more issues by quickly accessing the relevant lines of your source code.
```
Is "slow profiles, and more issues" correct? It doesn't read naturally. Perhaps rephrase?

Reviewer: Disregard this, reading it again, it does make sense. Although personally, "other" sounds better to me than "more"

> Datadog's source code integration allows you to connect your telemetry with your Git repositories hosted in GitHub, GitLab, or Bitbucket. Once you have enabled the [GitHub integration][5], you can debug stack traces, slow profiles, and other issues by quickly accessing the relevant lines of your source code.

Author: ```suggestion
Datadog's source code integration allows you to connect your telemetry with your Git repositories hosted in GitHub, GitLab, or Bitbucket. Once you have enabled the [source code integration][7], you can debug stack traces, slow profiles, and other issues by quickly accessing the relevant lines of your source code. 
```

> {{< img src="integrations/guide/source_code_integration/inline-code-snippet.png" alt="Inline code snippet of a Java RuntimeException with a button to view the code in Github" style="width:100%;">}}

Author: Show dropdown menu with three options

> By setting up the GitHub integration, you can see inline code snippets in **Error Tracking for Backend Services** and the **Continuous Profiler**. For more information, see [Inline Source Code](#inline-source-code) and [Continuous Profiler](#continuous-profiler).

Author: ```suggestion
```

## Setup

Datadog Agent v7.35.0 or later is required.

If you have [APM][6] set up already, navigate to [**Integrations** > **Link Source Code**][7] and configure the source code integration for your backend services.

## Tag your telemetry

> To link data to a specific commit, tag your telemetry with `git.commit.sha` and `git.repository_url` tags. Ensure that the `git.repository_url` tag does not contain protocols. For example, if your repository RUL is `https://github.con/example_repo`, the value for the `git.repository_url` tag should be `github.com/example_repo`.

Author: ```suggestion
To link data to a specific commit, tag your telemetry with `git.commit.sha` and `git.repository_url` tags. Ensure that the `git.repository_url` tag does not contain protocols. For example, if your repository URL is `https://github.con/example_repo`, the value for the `git.repository_url` tag should be `github.com/example_repo`.
```

{{% /tab %}}
{{< /tabs >}}

Datadog only captures the repository URL, the commit SHA of the current branch, and a list of tracked file paths—Datadog does not ingest or store any user code.

## Configure your repositories

{{< tabs >}}
{{% tab "GitHub" %}}

If you are a GitHub SaaS user, install Datadog's [GitHub integration][1] on the [GitHub integration tile][2] to link your telemetry with your source code. When specifying permissions on the integration tile, enable Datadog read permissions to **Contents**.

[1]: https://docs.datadoghq.com/integrations/github/
[2]: https://app.datadoghq.com/integrations/github/
{{% /tab %}}
{{% tab "Other Git Providers" %}}

To link telemetry to your source code, Datadog collects metadata for every commit SHA from your Git repository with the [`datadog-ci git-metadata upload`][1] command.

> When you run `datadog-ci git-metadata upload` within a git repository, Datadog receives the repository URL, the commit SHA of the current branch, and a list of tracked file paths.

Author: ```suggestion
When you run `datadog-ci git-metadata upload` within a Git repository, Datadog receives the repository URL, the commit SHA of the current branch, and a list of tracked file paths.
```

{{% /tab %}}

> {{< /tabs >}}

Author: ```suggestion
{{< /tabs >}}

By setting up the GitHub integration, you can see inline code snippets in **Error Tracking**. For more information, see [Inline Source Code](#inline-source-code).
```

> ## Examples of links to Git repositories

Author: ```suggestion
## Benefits of source code integration
```

> ### Inline source code

Author: Call out Links to your repositories (whatever source code manager is) first and add steps 1 to 5. Callout Error Tracking and Continuous Profiler. links to git providers: error tracking, CP.

Then second section is Inline source code. error tracking with that snapshot, need the github integration for this to happen.

> If you are a GitHub SaaS user, install Datadog's [GitHub integration][2] to directly inline code snippets from your GitHub repository in your stack traces in [Error Tracking for Backend Services][8]. When specifying permissions on the integration tile, enable Datadog read permissions to **Contents**.

Author: ```suggestion
If you are a GitHub SaaS user, install Datadog's [GitHub integration][2] to directly inline code snippets from your GitHub repository in your stack traces in [Error Tracking][8]. When specifying permissions on the integration tile, enable Datadog read permissions to **Contents**.
```

1. Navigate to [**APM** > **Error Tracking**][1].
2. Click on an issue. The **Issue Details** panel appears on the right.
3. Click **Connect to Preview** and **Authorize** to access the source code snippet containing the error.
4. Under **Latest Event**, click the **View Code** button to the right of a frame containing the code snippet.

> 5. Click **View file on GitHub**, **View Git blame**, or **View commit** to be redirected to GitHub.

Author: ```suggestion
5. Click **View file**, **View Git blame**, or **View commit** to be redirected to GitHub.
```
